lib/class-wp-rest-search-controller.php

```diff
@@ -0,0 +1,539 @@
+			$parent = get_post( $post->post_parent );
+			if ( $parent ) {
+				return $this->check_read_permission( $parent );
+			}
+		}
+
+		/*
+		 * If there isn't a parent, but the status is set to inherit, assume
+		 * it's published (as per get_post_status()).
+		 */
+		if ( 'inherit' === $post->post_status ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Prepares a single search result for response.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @param WP_Post         $post    Post object.
+	 * @param WP_REST_Request $request Request object.
+	 * @return WP_REST_Response Response object.
+	 */
+	public function prepare_item_for_response( $post, $request ) {
+
+		$GLOBALS['post'] = $post;
+
+		setup_postdata( $post );
+
+		if ( method_exists( $this, 'get_fields_for_response' ) ) {
+			$fields = $this->get_fields_for_response( $request );
+		} else {
+			$schema = $this->get_item_schema();
+			$fields = array_keys( $schema['properties'] );
+		}
+
+		// Base fields for every post.
+		$data = array();
+
+		if ( in_array( 'id', $fields, true ) ) {
+			$data['id'] = $post->ID;
+		}
+
+		if ( in_array( 'type', $fields, true ) ) {
+			$data['type'] = $post->post_type;
+		}
+
+		if ( in_array( 'link', $fields, true ) ) {
+			$data['link'] = get_permalink( $post->ID );
+		}
+
+		if ( in_array( 'title', $fields, true ) ) {
+			if ( post_type_supports( $post->post_type, 'title' ) ) {
+				add_filter( 'protected_title_format', array( $this, 'protected_title_format' ) );
+
+				$data['title'] = array(
+					'raw'      => $post->post_title,
+					'rendered' => get_the_title( $post->ID ),
+				);
+
+				remove_filter( 'protected_title_format', array( $this, 'protected_title_format' ) );
+			} else {
+				$data['title'] = array(
+					'raw'      => '',
+					'rendered' => '',
+				);
+			}
+		}
+
+		$context = ! empty( $request['context'] ) ? $request['context'] : 'view';
+		$data    = $this->add_additional_fields_to_object( $data, $request );
+		$data    = $this->filter_response_by_context( $data, $context );
+
+		// Wrap the data in a response object.
+		$response = rest_ensure_response( $data );
+
+		$response->add_links( $this->prepare_links( $post ) );
+
+		return $response;
+	}
+
+	/**
+	 * Overwrites the default protected title format.
+	 *
+	 * By default, WordPress will show password protected posts with a title of
+	 * "Protected: %s". As the REST API communicates the protected status of a post
+	 * in a machine readable format, we remove the "Protected: " prefix.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @return string Protected title format.
+	 */
+	public function protected_title_format() {
+		return '%s';
+	}
+
+	/**
+	 * Retrieves the item schema, conforming to JSON Schema.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+
+		$schema = array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => 'search',
```

Comment:
Should this be `search-result`?

Reply:
I agree to remove handling of custom post statuses, however we need to support `inherit` for attachments. This has been done before and is nothing we need to think about how to achieve.

Reply:
Let me correct myself. Custom post statuses in the REST API are nothing new and it is clear how to handle them. #3144 deals with how to handle custom post statuses in Gutenberg, but this controller is not exclusive to Gutenberg. Before that ticket is solved, we can simply use pass `status=publish` (or `status=publish,inherit`) to the endpoint.